File: src/elixir/svd/lib/svd/models/fuel_and_bunker_information.ex
defmodule SVD.Models.FuelAndBunkerInformation do
  @moduledoc false

  @fields [
    :main_engine_fuel_oil_consumption,
    :main_engine_fuel_oil_type,
    :main_engine_fuel_oil_rob,
    :auxiliary_engine_fuel_oil_consumption,
    :auxiliary_engine_fuel_oil_type,
    :auxiliary_engine_fuel_oil_rob,
    :boiler_fuel_oil_consumption,
    :boiler_fuel_oil_type,
    :boiler_fuel_oil_rob,
    :incinerator_fuel_oil_consumption,
    :gas_fuel_consumption,
    :gas_fuel_type,
    :gas_fuel_rob,
    :mgo_consumption,
    :mgo_rob,
    :hfo_consumption,
    :hfo_rob,
    :mdo_consumption,
    :mdo_rob,
    :total_fuel_oil_bunkered,
    :fuel_type,
    :fuel_type_trade_name,
    :bunker_delivery_note_number,
    :bunker_delivery_date_time,
    :fuel_proof_of_sustainability_reference,
    :fuel_bunkered,
    :fuel_mass,
    :fuel_density,
    :fuel_sulphur_content,
    :fuel_viscosity,
    :fuel_water_content,
    :fuel_higher_heating_value,
    :fuel_lower_heating_value,
    :fuel_calorific_value_reporting_scheme_code,
    :fuel_lower_calorific_value,
    :fuel_grade,
    :fuel_ghg_intensity_imo_manual,
    :fuel_ghg_intensity_imo_voyage,
    :fuel_bunker_port,
    :fuel_bunker_port_name,
    :fuel_carbon_dioxide_emission,
    :total_fuel_consumed,
    :fuel_consumed_by_main_engine,
    :fuel_consumed_by_diesel_electric_propulsion,
    :fuel_consumed_by_diesel_generator,
    :fuel_consumed_by_auxiliary_boiler,
    :fuel_consumed_by_auxiliary_engine,
    :fuel_consumed_by_cargo_heating,
    :fuel_consumed_by_reefer_containers,
    :fuel_consumed_by_discharge_pump,
    :fuel_consumed_by_other_devices,
    :fuel_remaining_on_board,
    :sludge_remaining_on_board
  ]

  @derive {Jason.Encoder, only: @fields}
  defstruct @fields

  @type t :: %__MODULE__{}

  @spec fields() :: [atom()]
  def fields, do: @fields
end

File: src/go/svd/models/fuel_and_bunker_information.go
package models

import "time"

// FuelAndBunkerInformation represents the "Fuel and Bunker" section of SVD.
type FuelAndBunkerInformation struct {
	FuelType                               string    `json:"fuelType" xml:"FuelType" csv:"FuelType"`
	FuelTypeTradeName                      string    `json:"fuelTypeTradeName" xml:"FuelTypeTradeName" csv:"FuelTypeTradeName"`
	BunkerDeliveryNoteNumber               string    `json:"bunkerDeliveryNoteNumber" xml:"BunkerDeliveryNoteNumber" csv:"BunkerDeliveryNoteNumber"`
	BunkerDeliveryDateTime                 time.Time `json:"bunkerDeliveryDateTime" xml:"BunkerDeliveryDateTime" csv:"BunkerDeliveryDateTime"`
	FuelProofOfSustainabilityReference     string    `json:"fuelProofOfSustainabilityReference" xml:"FuelProofOfSustainabilityReference" csv:"FuelProofOfSustainabilityReference"`
	FuelBunkered                           float64   `json:"fuelBunkered" xml:"FuelBunkered" csv:"FuelBunkered"`
	FuelMass                               float64   `json:"fuelMass" xml:"FuelMass" csv:"FuelMass"`
	FuelDensity                            float64   `json:"fuelDensity" xml:"FuelDensity" csv:"FuelDensity"`
	FuelSulphurContent                     float64   `json:"fuelSulphurContent" xml:"FuelSulphurContent" csv:"FuelSulphurContent"`
	FuelViscosity                          float64   `json:"fuelViscosity" xml:"FuelViscosity" csv:"FuelViscosity"`
	FuelWaterContent                       float64   `json:"fuelWaterContent" xml:"FuelWaterContent" csv:"FuelWaterContent"`
	FuelHigherHeatingValue                 float64   `json:"fuelHigherHeatingValue" xml:"FuelHigherHeatingValue" csv:"FuelHigherHeatingValue"`
	FuelLowerHeatingValue                  float64   `json:"fuelLowerHeatingValue" xml:"FuelLowerHeatingValue" csv:"FuelLowerHeatingValue"`
	FuelCalorificValueReportingSchemeCode  string    `json:"fuelCalorificValueReportingSchemeCode" xml:"FuelCalorificValueReportingSchemeCode" csv:"FuelCalorificValueReportingSchemeCode"`
	FuelLowerCalorificValue                float64   `json:"fuelLowerCalorificValue" xml:"FuelLowerCalorificValue" csv:"FuelLowerCalorificValue"`
	FuelGrade                              string    `json:"fuelGrade" xml:"FuelGrade" csv:"FuelGrade"`
	FuelGHGIntensityIMOManual              float64   `json:"fuelGHGIntensityIMOManual" xml:"FuelGHGIntensityIMOManual" csv:"FuelGHGIntensityIMOManual"`
	FuelGHGIntensityIMOVoyage              float64   `json:"fuelGHGIntensityIMOVoyage" xml:"FuelGHGIntensityIMOVoyage" csv:"FuelGHGIntensityIMOVoyage"`
	FuelBunkerPort                         string    `json:"fuelBunkerPort" xml:"FuelBunkerPort" csv:"FuelBunkerPort"`
	FuelBunkerPortName                     string    `json:"fuelBunkerPortName" xml:"FuelBunkerPortName" csv:"FuelBunkerPortName"`
	FuelCarbonDioxideEmission              float64   `json:"fuelCarbonDioxideEmission" xml:"FuelCarbonDioxideEmission" csv:"FuelCarbonDioxideEmission"`
	TotalFuelConsumed                      float64   `json:"totalFuelConsumed" xml:"TotalFuelConsumed" csv:"TotalFuelConsumed"`
	FuelConsumedByMainEngine               float64   `json:"fuelConsumedByMainEngine" xml:"FuelConsumedByMainEngine" csv:"FuelConsumedByMainEngine"`
	FuelConsumedByDieselElectricPropulsion float64   `json:"fuelConsumedByDieselElectricPropulsion" xml:"FuelConsumedByDieselElectricPropulsion" csv:"FuelConsumedByDieselElectricPropulsion"`
	FuelConsumedByDieselGenerator          float64   `json:"fuelConsumedByDieselGenerator" xml:"FuelConsumedByDieselGenerator" csv:"FuelConsumedByDieselGenerator"`
	FuelConsumedByAuxiliaryBoiler          float64   `json:"fuelConsumedByAuxiliaryBoiler" xml:"FuelConsumedByAuxiliaryBoiler" csv:"FuelConsumedByAuxiliaryBoiler"`
	FuelConsumedByAuxiliaryEngine          float64   `json:"fuelConsumedByAuxiliaryEngine" xml:"FuelConsumedByAuxiliaryEngine" csv:"FuelConsumedByAuxiliaryEngine"`
	FuelConsumedByCargoHeating             float64   `json:"fuelConsumedByCargoHeating" xml:"FuelConsumedByCargoHeating" csv:"FuelConsumedByCargoHeating"`
	FuelConsumedByReeferContainers         float64   `json:"fuelConsumedByReeferContainers" xml:"FuelConsumedByReeferContainers" csv:"FuelConsumedByReeferContainers"`
	FuelConsumedByDischargePump            float64   `json:"fuelConsumedByDischargePump" xml:"FuelConsumedByDischargePump" csv:"FuelConsumedByDischargePump"`
	FuelConsumedByOtherDevices             float64   `json:"fuelConsumedByOtherDevices" xml:"FuelConsumedByOtherDevices" csv:"FuelConsumedByOtherDevices"`
	FuelRemainingOnBoard                   float64   `json:"fuelRemainingOnBoard" xml:"FuelRemainingOnBoard" csv:"FuelRemainingOnBoard"`
	SludgeRemainingOnBoard                 float64   `json:"sludgeRemainingOnBoard" xml:"SludgeRemainingOnBoard" csv:"SludgeRemainingOnBoard"`
}
